File: assignment 7/python/programs/4.py
class MultipleQueue:
    def __init__(self, num_queues, queue_size):
        self.num_queues = num_queues
        self.queue_size = queue_size
        self.array = [None] * (num_queues * queue_size)
        self.front = [-1] * num_queues
        self.rear = [-1] * num_queues

    def is_empty(self, queue_num):
        return self.front[queue_num] == -1

    def is_full(self, queue_num):
        return (self.rear[queue_num] + 1) % self.queue_size == self.front[queue_num]

    def enqueue(self, queue_num, item):
        if self.is_full(queue_num):
            print(f"Queue {queue_num} is full. Cannot enqueue item.")
        else:
            if self.is_empty(queue_num):
                self.front[queue_num] = self.rear[queue_num] = 0
            else:
                self.rear[queue_num] = (self.rear[queue_num] + 1) % self.queue_size
            self.array[queue_num * self.queue_size + self.rear[queue_num]] = item
            print(f"Enqueued in Queue {queue_num}: {item}")

    def dequeue(self, queue_num):
        if self.is_empty(queue_num):
            print(f"Queue {queue_num} is empty. Cannot dequeue.")
        else:
            item = self.array[queue_num * self.queue_size + self.front[queue_num]]
            if self.front[queue_num] == self.rear[queue_num]:
                self.front[queue_num] = self.rear[queue_num] = -1
            else:
                self.front[queue_num] = (self.front[queue_num] + 1) % self.queue_size
            print(f"Dequeued from Queue {queue_num}: {item}")
            return item

    def display(self, queue_num):
        if self.is_empty(queue_num):
            print(f"Queue {queue_num} is empty.")
        else:
            print(f"Queue {queue_num} elements:", end=" ")
            i = self.front[queue_num]
            while True:
                print(self.array[queue_num * self.queue_size + i], end=" ")
                if i == self.rear[queue_num]:
                    break
                i = (i + 1) % self.queue_size
            print()


# Example usage:
if __name__ == "__main__":
    num_queues = 3
    queue_size = 5
    multi_queue = MultipleQueue(num_queues, queue_size)

    multi_queue.enqueue(0, 10)
    multi_queue.enqueue(1, 20)
    multi_queue.enqueue(2, 30)
    multi_queue.enqueue(0, 40)
    multi_queue.enqueue(1, 50)
    multi_queue.enqueue(2, 60)
    multi_queue.display(0)
    multi_queue.display(1)
    multi_queue.display(2)

    multi_queue.dequeue(0)
    multi_queue.dequeue(1)
    multi_queue.dequeue(2)
    multi_queue.display(0)
    multi_queue.display(1)
    multi_queue.display(2)

File: assignment 7/c/programs/4.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100

int queue[MAX_SIZE];
int front[2] = {-1, MAX_SIZE};
int rear[2] = {-1, MAX_SIZE};

void enqueue(int element, int queueNumber)
{
   if (front[queueNumber] == rear[queueNumber] - 1)
   {
      printf("Queue overflow\n");
      return;
   }
   if (queueNumber == 0)
   {
      queue[++front[queueNumber]] = element;
   }
   else if (queueNumber == 1)
   {
      queue[--rear[queueNumber]] = element;
   }
   else
   {
      printf("Invalid queue number\n");
   }
}

int dequeue(int queueNumber)
{
   if (front[queueNumber] == -1 || front[queueNumber] > rear[queueNumber])
   {
      printf("Queue underflow\n");
      return -1;
   }
   if (queueNumber == 0)
   {
      return queue[front[queueNumber]--];
   }
   else if (queueNumber == 1)
   {
      return queue[rear[queueNumber]++];
   }
   else
   {
      printf("Invalid queue number\n");
      return -1;
   }
}

void display(int queueNumber)
{
   if (front[queueNumber] == -1 || front[queueNumber] > rear[queueNumber])
   {
      printf("Queue is empty\n");
      return;
   }
   printf("Queue %d elements are: ", queueNumber);
   if (queueNumber == 0)
   {
      for (int i = front[queueNumber]; i <= rear[queueNumber]; i++)
      {
         printf("%d ", queue[i]);
      }
   }
   else if (queueNumber == 1)
   {
      for (int i = rear[queueNumber]; i <= front[queueNumber]; i++)
      {
         printf("%d ", queue[i]);
      }
   }
   else
   {
      printf("Invalid queue number\n");
   }
}

int main()
{
   enqueue(10, 0);
   enqueue(20, 0);
   enqueue(30, 1);
   enqueue(40, 1);
   display(0);
   display(1);
   dequeue(0);
   dequeue(1);
   display(0);
   display(1);
}
